src/merge_operator.rs

```diff
@@ -451,4 +454,35 @@ mod test {
         }
         assert!(DB::destroy(&opts, path).is_ok());
     }
+
+    #[test]
+    fn failed_merge_test() {
+        fn test_failing_merge(
+            _key: &[u8],
+            _val: Option<&[u8]>,
+            _operands: &mut MergeOperands,
+        ) -> Option<Vec<u8>> {
+            None
+        }
+        use crate::{Options, DB};
+
+        let path = "_rust_rocksdb_failed_merge_test";
+        let mut opts = Options::default();
+        opts.create_if_missing(true);
+        opts.set_merge_operator("test operator", test_failing_merge, None);
+        {
+            let db = DB::open(&opts, path).expect("open with a merge operator");
+            db.put(b"key", b"value").expect("put_ok");
+            let res = db.merge(b"key", b"new value");
+            match res.and_then(|_e| db.get(b"key")) {
+                Ok(val) => panic!("expected merge failure to propagate, got: {:?}", val),
+                Err(e) => {
+                    assert!(e.into_string().contains("Could not perform merge."));
+                }
+            }
+            // We can't DB::destroy, because that will detect the
+            // corruption and fail.
+            assert!(fs::remove_dir_all(path).is_ok());
```

Comment:
Using `expect` or `unwrap` here would be better because we can see the reason of failure.

Does this test pass for you locally?

Reply:
Updated!

This is passing for me locally, yes.

Reply:
Same for me. 

At the first glance, `DestroyDB` [should](https://github.com/facebook/rocksdb/blob/master/db/db_impl/db_impl.cc#L3665) delete everything even if something went wrong and report the reason of the first failure. But as far as I can see, there are still some files remains in this case.

Could you please try calling `DB::destroy` (and ignore the return value) before calling the fs::remove_dir_all?

Reply:
Honestly, I have run out of ideas.

Perhaps, `TempDir` from the `tempfile` crate can be used. At least it generates a directory with a random name, so it won't affect consequent test runs and the temp directory should be removed later even in case of a locked file.